### PR TITLE
Bad use of double quotes in rsyslog_files_permissions/bash/shared.sh (second try)

### PR DESCRIPTION
**Pull request by brett060102**
In our testing the log files specified in /etc/rsyslog.d/ were not getting remediated.
The issue is with line:
    RSYSLOG_CONFIGS+=("${RSYSLOG_ETC_CONFIG}" "${RSYSLOG_INCLUDE_CONFIG[@]}" "${RSYSLOG_INCLUDE[@]}")
which produces a RSYSLOG_CONFIGS of
    /etc/rsyslog.conf /run/rsyslog/additional-log-sockets.conf /etc/rsyslog.d/*.conf /etc/rsyslog.d/*.frule

line:
    RSYSLOG_CONFIGS+=(${RSYSLOG_ETC_CONFIG} ${RSYSLOG_INCLUDE_CONFIG[@]} ${RSYSLOG_INCLUDE[@]})
poduces:
    /etc/rsyslog.conf /run/rsyslog/additional-log-sockets.conf /etc/rsyslog.d/21-cloudinit.conf /etc/rsyslog.d/remote.conf /etc/rsyslog.d/acpid.frule /etc/rsyslog.d/firewall.frule /etc/rsyslog.d/NetworkManager.frule

Use find to capture effected conf and frule files.

#### Description:

- recode config file handle to use find as opposed to shell expansion for getting list of config files.

#### Rationale:

- current *.conf and *.frule handling not working
-

### CODE REVIEW

File: linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh [hunk 1]
```diff
@@ -14,9 +14,17 @@ declare -a LOG_FILE_PATHS
...
 
-# Browse each file selected above as containing paths of log files
+# Get full list of files to be checked
 # ('/etc/rsyslog.conf' and '/etc/rsyslog.d/*.conf' in the default configuration)
-for LOG_FILE in "${RSYSLOG_CONFIGS[@]}"
+declare -a RSYSLOG_FILES
+for ENTRY in "${RSYSLOG_CONFIGS[@]}"
+do
+     mapfile -t FINDOUT < <(find "$(dirname ${ENTRY}\")" -maxdepth 1 -name "$(basename \"${ENTRY})")
+     RSYSLOG_FILES+=("${RSYSLOG_FILES[@]}" "${FINDOUT[@]}")
```
Comment (brett060102) [inline linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:23]: I don't know why the \" are needed, but they are; code does not work without them. I have tested the result on SLE15 and Ubuntu 20.04 and it works as expected on both. And passes spell check.

Comment (brett060102) [reply linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:23]: Code updated so comment above no longer applies.

**Comment by brett060102** [comment]
@yuumasato Can you please look at this and see if it looks like an acceptable solution to the issue in https://github.com/ComplianceAsCode/content/pull/8574 I does take care of the issue I was seeing and passes shellcheck.
Thank  you.
@brett060102

**Comment by brett060102** [comment]
@Mab879 Tank you.

**Review by yuumasato — CHANGES_REQUESTED** [review]
@brett060102 So the problem seems to be the expansion of globs in arrays, and double quoting the array causes the glob not to expand.

In RHEL, the default config file contains `include(file="/etc/rsyslog.d/*.conf" mode="optional")`, 
and due to the double quoting this gets in the for loop without being expanded.
`echo ${RSYSLOG_CONFIGS[@]}` and `echo "${RSYSLOG_CONFIGS[@]}"` expands differently.

With https://github.com/ComplianceAsCode/content/commit/33fa9020e7ee006df6c01812188c5e4335cd6c66#diff-b3a261983f580c0a14bfe622685710b716ddddb53814265f52bac6b39ffa0d67R15 the glob expansion got broken.

I think I see another issue in line 15, the `+=` causes the files to be added multiple times, changing it to `=` seemed to solved it for me. Do you also see files added multiple times?
` RSYSLOG_CONFIGS=("${RSYSLOG_ETC_CONFIG}" "${RSYSLOG_INCLUDE_CONFIG[@]}" "${RSYSLOG_INCLUDE[@]}")`

Could you just add a comment clarifying that `RSYSLOG_CONFIGS` may have globs and the added `for` loop adds the explicit file paths ?
I'm good with the proposed solution, pinging @matejak to double check.

This bug took some time to be found and fixed, could you add a test scenario so that this problem does not manifest again? We don't have any test scenario with a glob in the `include()` or `$IncludeConfig`.
I think one pass and fail test scenarios will suffice.

**Comment by brett060102** [comment]
Code changes are up. I can get the tests up tomorrow.

**Comment by brett060102** [comment]
@yuumasato test for wildcards in include are up.

**Review by yuumasato — APPROVED** [review]
LGTM, thank you very much for adding the test!

File: linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh [hunk 2]
```diff
@@ -14,9 +14,19 @@ declare -a LOG_FILE_PATHS
 declare -a RSYSLOG_CONFIGS
 RSYSLOG_CONFIGS+=("${RSYSLOG_ETC_CONFIG}" "${RSYSLOG_INCLUDE_CONFIG[@]}" "${RSYSLOG_INCLUDE[@]}")
```
Comment (yuumasato) [inline linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:15]: @brett060102 One more thing I noticed, I think the `+`  be dropped here too, do you agree?

File: linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh [hunk 3]
```diff
@@ -14,9 +14,19 @@ declare -a LOG_FILE_PATHS
...
-for LOG_FILE in "${RSYSLOG_CONFIGS[@]}"
+# Get full list of files to be checked
+# RSYSLOG_CONFIGS may contain globs such as 
+# /etc/rsyslog.d/*.conf /etc/rsyslog.d/*.frule
+# So, loop over the entries in RSYSLOG_CONFIGS and use find to get the list of included files.
+declare -a RSYSLOG_FILES
+for ENTRY in "${RSYSLOG_CONFIGS[@]}"
+do
+     mapfile -t FINDOUT < <(find "$(dirname "${ENTRY}")" -maxdepth 1 -name "$(basename "${ENTRY}")")
+     RSYSLOG_FILES=("${RSYSLOG_FILES[@]}" "${FINDOUT[@]}")
```
Comment (matejak) [inline linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:25]:
```suggestion
     RSYSLOG_FILES+=("${FINDOUT[@]}")
```

I see that in the PR description there were mentions of duplicates when using `+=`, but that was probably caused by missing initialization of the array, so an earlier remediation influenced its contents. This proposed change should result in an exactly the same behavior, could you please comment on that?

File: linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh [hunk 4]
```diff
@@ -14,9 +14,19 @@ declare -a LOG_FILE_PATHS
...
 RSYSLOG_CONFIGS+=("${RSYSLOG_ETC_CONFIG}" "${RSYSLOG_INCLUDE_CONFIG[@]}" "${RSYSLOG_INCLUDE[@]}")
 
-# Browse each file selected above as containing paths of log files
-# ('/etc/rsyslog.conf' and '/etc/rsyslog.d/*.conf' in the default configuration)
-for LOG_FILE in "${RSYSLOG_CONFIGS[@]}"
+# Get full list of files to be checked
+# RSYSLOG_CONFIGS may contain globs such as 
+# /etc/rsyslog.d/*.conf /etc/rsyslog.d/*.frule
+# So, loop over the entries in RSYSLOG_CONFIGS and use find to get the list of included files.
+declare -a RSYSLOG_FILES
```
Comment (matejak) [inline linux_os/guide/system/logging/ensure_rsyslog_log_file_configuration/rsyslog_files_permissions/bash/shared.sh:21]:
```suggestion
RSYSLOG_FILES=()
```

Declaration doesn't empty the variable if it already exists. What about initializing it right away with an empty array?

**Comment by matejak** [comment]
> I think I see another issue in line 15, the += causes the files to be added multiple times, changing it to = seemed to solved it for me. Do you also see files added multiple times?

The commit didn't change the behavior, but it is probably true that the previous behavior was wrong already. I would blame the missing initialization.

I am not sure about expanding the wildcard if there is an option to have a literal asterisk in the configuration - the remediation-time expansion "freezes" the configuration in time, so when one adds another config file to that directory, it won't work, although other config files next to it work.

**Review by brett060102 — COMMENTED** [review]
Accept review suggestions, more efficient  handling of RSYSLOG_FILES

**Comment by brett060102** [comment]
@matejak The code updates have been made. I am not sure I understand the comment:
"I am not sure about expanding the wildcard if there is an option to have a literal asterisk in the configuration - the remediation-time expansion "freezes" the configuration in time, so when one adds another config file to that directory, it won't work, although other config files next to it work."

Ubuntu has:
/etc/rsyslog.conf
$IncludeConfig /etc/rsyslog.d/*.conf
SUSE has:
$IncludeConfig /etc/rsyslog.d/*.conf
$IncludeConfig /etc/rsyslog.d/*.frule

So, the wildcards are normal.

The rough flow here is: check /etc/rsyslog.conf  for includes. If includes are wildcards, then expand the wildcards to get a full list of files that define rsyslog log files. Those are then processed to get a list of logfiles that are checked to make sure they have then right privs set.

But this is all done when remediation is run and not when remediation is built. So, the list of log files to check is determined when remediation is run. If a new config which defines a new log file is added after remediation is initially run,  then the new log file would not be checked and remediated until the next remediation run.

Am I missing your point here?
Thanks.

**Comment by matejak** [comment]
> ... Those are then processed to get a list of logfiles that are checked to make sure they have then right privs set.

Sorry, I have misunderstood how the remediation works. Thanks for clarifying it, I have no further objections.

**Pull request merged**